thewalrus/quantum/photon_number_distributions.py

```diff
@@ -89,3 +90,90 @@ def _convolve_squeezed_state_distribution(s, cutoff=50, padding_factor=2):
     for s_val in s:
         ps = np.convolve(ps, _squeezed_state_distribution(s_val, cutoff_sc))[0:cutoff_sc]
     return ps
+
+
+def total_photon_number_distribution(n, k, s, eta, pref=1.0):
+    r"""Probability of observing a total of :math:`n` photons when :math:`k` identical
+    single-mode squeezed vacua with squeezing parameter s undergo loss by transmission :math:`\eta`.
+
+    For the derivation see Appendix E of `'Quantum Computational Supremacy via High-Dimensional Gaussian Boson Sampling',
+    Deshpande et al. <https://arxiv.org/abs/2102.12474>`_.
+
+
+    Args:
+        n (int): number of photons
+        k (int): number of squeezed modes
+        s (float): squeezing parameter
+        eta (float): transmission parameter, between 0 and 1 inclusive
+        pref (float): use to return the probability times pref**n
+    Returns:
+        (float): probability of observing a total of n photons or the probability times `pref ** n`.
+    """
+    if n % 2 == 0:
+        peven = (
+            binom(-1 + k / 2.0 + n / 2.0, n / 2.0)
+            * hyp2f1(0.5 + n / 2.0, k / 2.0 + n / 2.0, 0.5, (1 - eta) ** 2 * np.tanh(s) ** 2)
+            * (1 / np.cosh(s)) ** k
+            * (pref * eta * np.tanh(s)) ** n
```

Comment:
I'm not sure how this `pref` comes in here, but I assume it's correct. It was the only part I didn't recognize in the equation. 😆

Reply:
I see. So, you at least agree that if `pref=1` everything checks. Now the idea is that we want to calculate the probability times pref**n. If you check both the even and the odd cases you can verify that this indeed the case. I just have put the prefactor `pref` in the same bracket as any other terms that is also raised to the power `n`, namely `eta * np.tanh(s)` to make it more stable. For example imagine that prefactor is `pref=1.99` and `eta*np.tanh(s) = 0.5` then if you take power separately you will find a gigantic number times a very small number, if you multiply them together and then take the power you will get something like `0.995*n` which is a lot nicer numerically.